src/Agent.Plugins/PipelineArtifact/PipelineArtifactServer.cs

```diff
@@ -313,10 +324,17 @@ public class PipelineArtifactServer
             }
         }
 
-        private CallbackAppTraceSource CreateTracer(AgentTaskPluginExecutionContext context)
-        {
-            var tracer = new CallbackAppTraceSource(str => context.Output(str), System.Diagnostics.SourceLevels.Information);
-            return tracer;
+        public class Tracer {
+            private static CallbackAppTraceSource tracer;
```

Comment:
This seems overkill,

I would make Tracer a readonly member variable of PipelineArtifactServer so when you construct PipelineArtifactServer, you will need to pass the AgentContext as part of the constructor which will then fill in the Tracer variable. That way you only have one Tracer in the PipelineArtifactServer which can be recycled across multiple calls.